File: hashmap/group-anagram/group_anagram.py
# https://leetcode.com/problems/group-anagrams/?envType=study-plan-v2&envId=top-interview-150
# Time complexity: O(N * K) where N is the number of words and k is the number of characters in the words
# Space complexity: O(N * K) where N is the number of words and k is the number of characters in the words

from collections import Counter

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        words_by_character_counts = defaultdict(list)

        for word in strs:
            curr_char_count = self.calculate_character_count(word)
            words_by_character_counts[curr_char_count].append(word)

        return list(words_by_character_counts.values())

    def calculate_character_count(self, word: str):
        lst = [0] * 26
        
        for letter in word:
            index = ord(letter) - ord('a')
            lst[index] += 1

        return tuple(lst)

File: hashmap/group-anagram/GroupAnagram.java
// https://leetcode.com/problems/group-anagrams/?envType=study-plan-v2&envId=top-interview-150

// Time complexity: O(N.K)
// Space Complexity: O(N · K) due to storing grouped strings and keys

import java.util.*;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {

        HashMap<String, List<String>> wordsByCharCount= new HashMap<>();

        for(String str: strs) {
            String charCountMap = getCharCountMap(str);

            if(wordsByCharCount.containsKey(charCountMap)) {
                wordsByCharCount.get(charCountMap).add(str);
            } else {
                var newList = new ArrayList<>(List.of(str));
                wordsByCharCount.put(charCountMap, newList);
            }
        }

        return new ArrayList(wordsByCharCount.values());
    }

    private String getCharCountMap(String str) {

        int[] charCount = new int[26];
        int size = str.length();
        int curr = 0;

        while(curr < size) {
            char c = str.charAt(curr);
            int index = (int) c - 'a';
            charCount[index] = charCount[index] + 1;
            curr++;
        }

        return Arrays.toString(charCount);
    }
}